File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第10章补给.py
#补给 lanqiaoOJ  1050 

from math import *
def dis(i, j):
    return sqrt((xy[i][0]-xy[j][0])**2+(xy[i][1]-xy[j][1])**2)
def floyd():
    global w
    for k in range(n):
        for i in range(n):
            for j in range(n):
                w[i][j]=min(w[i][j],w[i][k]+w[k][j])
n,D = map(int,input().split())
dp = [[float('inf')]*(21) for _ in range((1<<21))]
w = [[float('inf')]*(21) for _ in range(21)]
xy = []  #坐标
for _ in range(n): xy.append(list(map(float, input().split())))
for i in range(n):
    for j in range(i+1,n):
        w[i][j] =dis(i,j); w[j][i]=w[i][j]
        if w[i][j]>D: w[i][j]=float('inf');w[j][i]=float('inf')
floyd()
for i in range(i<(1<<n)):
    for j in range(n):    dp[i][j]=float('inf')
dp[1][0] = 0
for S in range(1<<n):
    for j in range(n):
        if (S>>j) & 1:
           for k in range(n):
               if (S^(1<<j)) >> k & 1 :
                  dp[S][j] = min(dp[S][j],dp[S^(1<<j)][k] + w[k][j])
ans = inf
for i in range(1,n):   ans=min(ans,w[i][0]+dp[(1<<n)-1][i])
print("%.2f" % ans)         

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第10章补给.java
//补给 lanqiaoOJ 1050 

import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static Scanner cin = new Scanner(System.in);
    static int INF = 0x3f3f3f3f;
    static double[] x = new double[20];
    static double[] y = new double[20];
    static int n;
    static double D;
    static double[][] w = new double[20][20];
    static int[][] dp = new int[1 << 20][20];

    public static double dis(int a, int b) {
        return Math.sqrt((x[a] - x[b]) * (x[a] - x[b]) + (y[a] - y[b]) * (y[a] - y[b]));
    }

    public static void floyd() {
        for (int k = 0; k < n; k++)
            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++)
                    w[i][j] = Math.min(w[i][j], w[i][k] + w[k][j]);
    }

    public static void main(String[] args) {
        n = cin.nextInt();
        D = cin.nextDouble();
        for (int i = 0; i < n; i++) {
            x[i] = cin.nextDouble();
            y[i] = cin.nextDouble();
        }
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                w[i][j] = w[j][i] = dis(i, j);
                if (w[i][j] > D) w[i][j] = w[j][i] = INF;
            }
        }
        floyd();
        for (int i = 0; i < (1 << 20); i++)for(int j = 0 ; j < 20 ; j ++) dp[i][j] = INF;
        dp[1][0] = 0;
        for (int S = 0; S < (1 << n); S++) {
            for (int j = 0; j < n; j++) {
                if ((S >> j & 1) != 0) {
                    for (int k = 0; k < n; k++) {
                        if (((S ^ (1 << j)) >> k & 1) != 0)
                            dp[S][j] = Math.min(dp[S][j], dp[S ^ (1 << j)][k] + (int) (5000 * w[k][j]));
                    }
                }
            }
        }
        int ans =  INF;
        for (int i = 1; i < n; i++)
            ans = Math.min(ans, (int) (w[i][0] * 5000) + dp[(1 << n) - 1][i]);
        System.out.printf("%.2f%n", ans / 5000.0);
    }
}